File: 2-1/sub_sum.ml
let n = 4

let a = [1; 2; 4; 7]

let k = 15

let rec solve i sum =
  if i = n then (sum == k)
  else if ( solve (i+1) sum ) then true
  else if ( solve (i+1) (sum + (List.nth a i)) ) then true
  else false

let ans = solve 0 0

File: 2-1/sub_sum.cpp
#include <bits/stdc++.h>
using namespace std;

bool solve (vector<int> a, int k, int i, int sum) {
  int n = a.size();
  
  // n個までたどり着いた
  if ( i==n ) return sum == k;

  // n番目を使わない場合
  else if ( solve(a, k, i+1, sum) ) return true;

  // n番目を使う場合
  else if ( solve(a, k, i+1, sum+a.at(i)) ) return true;

  else return false;
}

int main() {
  int n = 4;
  vector<int> a = {1, 2, 4, 7};
  int k = 15;

  bool b = solve(a, k, 0, 0);
  if (b) cout << "Yes\n";
  else cout << "No\n";
}
